nfs/src/rpc_task.cpp

```diff
@@ -476,6 +476,377 @@ void rpc_task::run_setattr()
+
+    const char* errstr;
+    auto res = (READ3res*)data;
+    const int status = (task->status(rpc_status, NFS_STATUS(res), &errstr));
+    auto ino = task->rpc_api.read_task.get_inode();
+    auto readfile_handle =
+        task->get_client()->get_nfs_inode_from_ino(ino)->filecache_handle;
+
+    AZLogDebug("readfile_callback:: Bytes read: {} eof: {}, "
+               "requested_bytes: {} off: {}",
+               res->READ3res_u.resok.count, res->READ3res_u.resok.eof,
+               bc->length,
+               bc->offset);
+
+    // We should never get more data than what we requested.
+    assert (bc->length >= res->READ3res_u.resok.count);
+
+    if (status == 0) {
+        if (bc->is_empty && (bc->length == res->READ3res_u.resok.count))
+        {
+            /*
+             * Only the first read which got hold of the complete membuf
+             * will have this byte_chunk set to empty.
+             * Only such reads should set the uptodate flag.
+             * Also the uptodate flag should be set only if we have read
+             * the entire membuf.
+             */
+            AZLogDebug("Setting uptodate flag. offset: {}, length: {}",
+                       task->rpc_api.read_task.get_offset(),
+                       task->rpc_api.read_task.get_size());
+
+            bc->get_membuf()->set_uptodate();
+        }
+    }
+    else
+    {
+        assert(res->READ3res_u.resok.count == 0);
+
+        AZLogError("Read failed. Error: {} offset: {} size: {}",
+                   errstr,
+                   bc->offset,
+                   bc->length);
+    }
+
+    /*
+     * Release the lock that we held on the membuf since the data is now
+     * written to it.
+     * The lock is needed only to write the data and not to just read it.
+     * Hence it is safe to read this membuf even beyond this point.
+     */
+    bc->get_membuf()->clear_locked();
+
+    /*
+     * Since we come here only for client reads, we will not cache the date.
+     * Hence release the chunk.
+     * This can safely be done for both success and failure case.
+     */
+    readfile_handle->release(bc->offset, bc->length);
+
+    // Update the byte chunk length.
+    bc->length = res->READ3res_u.resok.count;
+
+    /*
+     * Take a lock here since multiple readfiles issued can try
+     * modifying the below members.
+     */
+    {
+        std::unique_lock<std::shared_mutex> lock(task->read_task_lock);
+
+        // Decrement the number of reads issued.
+        task->num_of_reads_issued_to_backend--;
+
+        if (task->readfile_completed) {
+            /*
+             * If readfile_completed is set, it means that there was a previous
+             * failure encountered as a result of which we have already sent
+             * failure response to the caller.
+             * Hence do not do anything here and just exit.
+             */
+            return;
+        }
+
+        if (status || (task->num_of_reads_issued_to_backend == 0))
```

Comment:
for status!=0 case, if we send the response we will free the rpc_task, so when other reads complete we won't have the task pointer valid.